### PR TITLE
Fold expressions with constan on LHS and binary sum or diff with constant on RHS

### PR DESCRIPTION
**Pull request by MaxGraey**
`C1 - (x + C2)`  ->  `(C1 - C2) - x`
`C1 - (x - C2)`  ->  `(C1 + C2) - x`
`C1 - (C2 - x)`  ->  `x + (C1 - C2)`

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 16051
   CompareVMs: 2870
   CheckDeterminism: 1397
   Wasm2JS: 1623
   Asyncify: 2681

ITERATION: 17739
```

**Comment by MaxGraey** [comment]
@tlively could you take a look at this? Thanks!

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -1733,6 +1733,40 @@ struct OptimizeInstructions
...
+          left->value = c1->value.add(c2->value);
+          curr->right = x;
+          return curr;
+        }
+      }
+      // C1 - (C2 - x)  ==>   x + (C1 - C2)
+      if (matches(curr,
+                  binary(Sub, ival(&c1), binary(Sub, ival(&c2), any(&x))))) {
+        left->value = c1->value.sub(c2->value);
+        if (left->value.isNegative()) {
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:1759]: Can you write out an extra comment illustrating this case? Something like 

```
// -C1 - (C2 - x)  ==>  x - (C1 - C2)`
```

**Review by tlively — COMMENTED** [review]
Generally looks good!

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -1733,6 +1733,40 @@ struct OptimizeInstructions
         !effects(curr->right).hasSideEffects()) {
       return curr->left;
     }
+    {
+      // C1 - (x + C2)  ==>  (C1 - C2) - x
+      // C1 - (x - C2)  ==>  (C1 + C2) - x
```
[the remark this answers is not in the document]
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp]: Sure! Done!

File: src/passes/OptimizeInstructions.cpp [hunk 3]
```diff
@@ -1733,6 +1733,39 @@ struct OptimizeInstructions
...
+        left->value = c1->value.add(c2->value);
+        curr->right = x;
+        return curr;
+      }
+      // C1 - (C2 - x)  ==>   x + (C1 - C2)
+      if (matches(curr,
+                  binary(Sub, ival(&c1), binary(Sub, ival(&c2), any(&x))))) {
+        left->value = c1->value.sub(c2->value);
+        if (left->value.isNegative()) {
+          // -C1 - (C2 - x)  ==>  x - (C1 - C2)`
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:1758]: Oops, looks my typo made it in 😅 
```suggestion
          // -C1 - (C2 - x)  ==>  x - (C1 - C2)
```

**Review by tlively — APPROVED** [review]

**Pull request merged**